File: app/src/main/java/ai/rotor/rotorvehicle/rotor_ctl/Arduino.java
package ai.rotor.rotorvehicle.rotor_ctl;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbManager;

import com.felhr.usbserial.UsbSerialDevice;
import com.felhr.usbserial.UsbSerialInterface;

import java.util.HashMap;

import static ai.rotor.rotorvehicle.RotorUtils.*;
import timber.log.Timber;

public class Arduino implements UsbSerialInterface.UsbReadCallback {
    private Context mContext;
    private UsbReceiver mUsbReceiver;
    private UsbManager mUsbManager;
    private boolean mIsOpened;
    private ArduinoListener mListener;
    private UsbDevice mLastArduinoAttached;
    private UsbSerialDevice mSerialPort;
    private UsbDeviceConnection mSerialConnection;
    private StringBuilder mBufferedChars = new StringBuilder();
    final private static String ACTION_USB_DEVICE_PERMISSION = "com.example.androidusb.USB_PERMISSION";


    public Arduino(Context context) {
        this.mContext = context;
        this.mUsbReceiver = new UsbReceiver();
        this.mUsbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
        this.mIsOpened = false;
    }

    public void setArduinoListener(ArduinoListener listener) {
        this.mListener = listener;

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(UsbManager.ACTION_USB_DEVICE_ATTACHED);
        intentFilter.addAction(UsbManager.ACTION_USB_DEVICE_DETACHED);
        intentFilter.addAction(ACTION_USB_DEVICE_PERMISSION);

        mContext.registerReceiver(mUsbReceiver, intentFilter);

        this.mLastArduinoAttached = getAttachedArduino();
        if (mLastArduinoAttached != null && listener != null) {
            listener.onArduinoAttached(mLastArduinoAttached);
        }
    }

    public void unSetArduinoListener() {
        this.mListener = null;
    }

    public void open(UsbDevice device) {
        PendingIntent permissionIntent = PendingIntent.getBroadcast(mContext, 0, new Intent(ACTION_USB_DEVICE_PERMISSION), 0);
        mUsbManager.requestPermission(device, permissionIntent);
    }

    public void reOpen() {
        open(mLastArduinoAttached);
    }

    public void close() {
        if (mSerialPort != null) {
            mSerialPort.close();
        }
        if (mSerialConnection != null) {
            mSerialConnection.close();
        }

        mIsOpened = false;
        mContext.unregisterReceiver(mUsbReceiver);
    }

    public void send(byte[] bytes) {
        if (mSerialPort != null) {
            mSerialPort.write(bytes);
        }
    }

    private class UsbReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            UsbDevice device;
            if (intent.getAction() != null) {
                switch (intent.getAction()) {
                    case UsbManager.ACTION_USB_DEVICE_ATTACHED:
                        device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
                        if (device.getVendorId() == ARDUINO_VENDOR_ID) {
                            mLastArduinoAttached = device;
                            if (mListener != null) {
                                mListener.onArduinoAttached(device);
                            }
                        }
                        break;
                    case UsbManager.ACTION_USB_DEVICE_DETACHED:
                        device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
                        if (device.getVendorId() == ARDUINO_VENDOR_ID) {
                            if (mListener != null) {
                                mListener.onArduinoDetached();
                            }
                        }
                        break;
                    case ACTION_USB_DEVICE_PERMISSION:
                        boolean permissionGranted = intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false);
                        if (permissionGranted) {
                            device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
                            if (device.getVendorId() == ARDUINO_VENDOR_ID) {
                                mSerialConnection = mUsbManager.openDevice(device);
                                mSerialPort = UsbSerialDevice.createUsbSerialDevice(device, mSerialConnection);
                                if (mSerialPort != null) {
                                    if (mSerialPort.open()) {
                                        mSerialPort.setBaudRate(BAUD_RATE);
                                        mSerialPort.setDataBits(UsbSerialInterface.DATA_BITS_8);
                                        mSerialPort.setStopBits(UsbSerialInterface.STOP_BITS_1);
                                        mSerialPort.setParity(UsbSerialInterface.PARITY_NONE);
                                        mSerialPort.setFlowControl(UsbSerialInterface.FLOW_CONTROL_OFF);
                                        mSerialPort.read(Arduino.this);

                                        mIsOpened = true;

                                        if (mListener != null) {
                                            mListener.onArduinoOpened();
                                        }
                                    }
                                } else {
                                    Timber.d("Serial port is null");
                                }
                            }
                        } else if (mListener != null) {
                            mListener.onUsbPermissionDenied();
                        }
                        break;
                }
            }
        }
    }

    private UsbDevice getAttachedArduino() {
        HashMap<String, UsbDevice> map = mUsbManager.getDeviceList();
        for (UsbDevice device : map.values()) {
            if (device.getVendorId() == ARDUINO_VENDOR_ID) {
                return device;
            }
        }
        return null;
    }

    public boolean isOpened() {
        return mIsOpened;
    }

    @Override
    public void onReceivedData(byte[] bytes) {
        String message = new String(bytes);
        if (mListener != null && message.length() > 0) {
            for (int i = 0; i < message.length(); i++) {
                char c = message.charAt(i);
                if (String.valueOf(c).matches(".")) {
                    mBufferedChars.append(c);

                } else {
                    mListener.onArduinoMessage(mBufferedChars.toString());
                    mBufferedChars.setLength(0);
                }
            }
        }
    }
}

File: app/src/main/java/ai/rotor/rotorvehicle/ctl/Arduino.kt
package ai.rotor.rotorvehicle.ctl

import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbDeviceConnection
import android.hardware.usb.UsbManager

import com.felhr.usbserial.UsbSerialDevice
import com.felhr.usbserial.UsbSerialInterface

import java.util.HashMap

import ai.rotor.rotorvehicle.RotorUtils
import timber.log.Timber

class Arduino internal constructor(private val mContext: Context) : UsbSerialInterface.UsbReadCallback {
    private val mUsbReceiver: UsbReceiver
    private val mUsbManager: UsbManager
    private var mIsOpened: Boolean = false
    private var mListener: ArduinoListener? = null
    private var mLastArduinoAttached: UsbDevice? = null
    private var mSerialPort: UsbSerialDevice? = null
    private var mSerialConnection: UsbDeviceConnection? = null
    private val mBufferedChars = StringBuilder()

    private val attachedArduino: UsbDevice?
        get() {
            val map = mUsbManager.deviceList
            for (device in map.values) {
                if (device.vendorId == RotorUtils.ARDUINO_VENDOR_ID) {
                    return device
                }
            }
            return null
        }


    init {
        this.mUsbReceiver = UsbReceiver()
        this.mUsbManager = mContext.getSystemService(Context.USB_SERVICE) as UsbManager
        this.mIsOpened = false
    }

    fun setArduinoListener(listener: ArduinoListener?) {
        this.mListener = listener

        val intentFilter = IntentFilter()
        intentFilter.addAction(UsbManager.ACTION_USB_DEVICE_ATTACHED)
        intentFilter.addAction(UsbManager.ACTION_USB_DEVICE_DETACHED)
        intentFilter.addAction(ACTION_USB_DEVICE_PERMISSION)

        mContext.registerReceiver(mUsbReceiver, intentFilter)

        this.mLastArduinoAttached = attachedArduino
        if (mLastArduinoAttached != null && listener != null) {
            listener.onArduinoAttached(mLastArduinoAttached!!)
        }
    }

    fun unSetArduinoListener() {
        this.mListener = null
    }

    fun open(device: UsbDevice?) {
        val permissionIntent = PendingIntent.getBroadcast(mContext, 0, Intent(ACTION_USB_DEVICE_PERMISSION), 0)
        mUsbManager.requestPermission(device, permissionIntent)
    }

    fun reOpen() {
        open(mLastArduinoAttached)
    }

    fun close() {
        if (mSerialPort != null) {
            mSerialPort!!.close()
        }
        if (mSerialConnection != null) {
            mSerialConnection!!.close()
        }

        mIsOpened = false
        mContext.unregisterReceiver(mUsbReceiver)
    }

    fun send(bytes: ByteArray) {
        if (mSerialPort != null) {
            mSerialPort!!.write(bytes)
        }
    }

    private inner class UsbReceiver : BroadcastReceiver() {

        override fun onReceive(context: Context, intent: Intent) {
            val device: UsbDevice
            if (intent.action != null) {
                when (intent.action) {
                    UsbManager.ACTION_USB_DEVICE_ATTACHED -> {
                        device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE)
                        if (device.vendorId == RotorUtils.ARDUINO_VENDOR_ID) {
                            mLastArduinoAttached = device
                            if (mListener != null) {
                                mListener!!.onArduinoAttached(device)
                            }
                        }
                    }
                    UsbManager.ACTION_USB_DEVICE_DETACHED -> {
                        device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE)
                        if (device.vendorId == RotorUtils.ARDUINO_VENDOR_ID) {
                            if (mListener != null) {
                                mListener!!.onArduinoDetached()
                            }
                        }
                    }
                    ACTION_USB_DEVICE_PERMISSION -> {
                        val permissionGranted = intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)
                        if (permissionGranted) {
                            device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE)
                            if (device.vendorId == RotorUtils.ARDUINO_VENDOR_ID) {
                                mSerialConnection = mUsbManager.openDevice(device)
                                mSerialPort = UsbSerialDevice.createUsbSerialDevice(device, mSerialConnection)
                                if (mSerialPort != null) {
                                    if (mSerialPort!!.open()) {
                                        mSerialPort!!.setBaudRate(RotorUtils.BAUD_RATE)
                                        mSerialPort!!.setDataBits(UsbSerialInterface.DATA_BITS_8)
                                        mSerialPort!!.setStopBits(UsbSerialInterface.STOP_BITS_1)
                                        mSerialPort!!.setParity(UsbSerialInterface.PARITY_NONE)
                                        mSerialPort!!.setFlowControl(UsbSerialInterface.FLOW_CONTROL_OFF)
                                        mSerialPort!!.read(this@Arduino)

                                        mIsOpened = true

                                        if (mListener != null) {
                                            mListener!!.onArduinoOpened()
                                        }
                                    }
                                } else {
                                    Timber.d("Serial port is null")
                                }
                            }
                        } else if (mListener != null) {
                            mListener!!.onUsbPermissionDenied()
                        }
                    }
                }
            }
        }
    }

    override fun onReceivedData(bytes: ByteArray) {
        val message = String(bytes)
        if (mListener != null && message.length > 0) {
            for (i in 0 until message.length) {
                val c = message[i]
                if (c.toString().matches(".".toRegex())) {
                    mBufferedChars.append(c)

                } else {
                    mListener!!.onArduinoMessage(mBufferedChars.toString())
                    mBufferedChars.setLength(0)
                }
            }
        }
    }

    companion object {
        private val ACTION_USB_DEVICE_PERMISSION = "com.example.androidusb.USB_PERMISSION"
    }
}
